_source/_posts/2022-04-11-java-cli-device-grant.adoc

```diff
@@ -0,0 +1,412 @@
+        .POST(formBody(authArgs))
+        .header("Accept", "application/json")
+        .header("Content-Type", "application/x-www-form-urlencoded")
+        .build();
+
+// execute the request
+HttpResponse<Map> httpResponse = client.send(request, rs -> parseJson(Map.class));
+
+// Fail on anything except a 200 response
+if (httpResponse.statusCode() != 200) {
+    System.err.println("Error: " + httpResponse);
+    throw new RuntimeException("Failed to initialize device grant authorization, likely " +
+                               "caused by an invalid IdP configuration");
+}
+
+Map authorizationResponse = httpResponse.body();
+
+String deviceCode = (String) authorizationResponse.get("device_code");
+String verificationUri = (String) authorizationResponse.get("verification_uri");
+String userCode = (String) authorizationResponse.get("user_code");
+String verificationUriComplete = (String) authorizationResponse.get("verification_uri_complete");
+Duration interval = Duration.ofSeconds((int) authorizationResponse.get("interval"));
+Duration expiresIn = Duration.ofSeconds((int) authorizationResponse.get("expires_in"));
+----
+
+NOTE: To keep the steps minimal in this post, I'm parsing the JSON to a map, but you could create a typed object to avoid any of the above casting.
+
+Assuming all went well, the response returned from the IdP will be a JSON block (parsed into the above values) that looks like this:
+
+====
+[source,json]
+----
+{
+  "device_code": "a-random-device-code", # <.>
+  "user_code": "ABCD1234", # <.>
+  "verification_uri": "https://id.bdemers.io/activate", # <.>
+  "verification_uri_complete": "https:/dev-133337/activate?user_code=ABCD1234", # <.>
+  "expires_in": 600, # <.>
+  "interval": 5 # <.>
+}
+----
+
+<.> `device_code` - A secret known by the device/application, it will be used in the following steps.
+<.> `user_code` - Shown to the user, and ties a different browser session to this device/application.
+<.> `verification_uri`- URL the user needs to visit and type enter the `user_code` before logging in.
+<.> `verification_uri_complete` - A URL combining the `verification_uri` and `user_code` which can be shown directly to the user (if they can copy and paste) or rendered as a QR code (on a TV).
+<.> `expires_in` - Number of seconds the user has to finish the flow before the codes expire.
+<.> `interval` - Minimum frequency in seconds the application should wait while polling the server to see if the user has been authorized.
+====
+
+=== User logs in via a browser
+
+The end user needs to open a browser to log in, how they do that is up to your application. For the best user experience, you may want to open a browser automatically. However, that might not be possible, so always fall back to showing the URL and code.
+
+[source,java]
+----
+// check if a browser can be opened
+if (!GraphicsEnvironment.isHeadless() && Desktop.getDesktop().isSupported(Desktop.Action.BROWSE)) {
```

Comment:
may be good time to mention we are using AWT here and may be even a line on what AWT is, I doubt devs who have never built Desktop java apps (looks at mirror) might not know what it is :wink: